x/upgrade/internal/keeper/keeper.go

```diff
@@ -0,0 +1,119 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/exported"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/internal/types"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+type Keeper struct {
+	storeKey        sdk.StoreKey
+	cdc             *codec.Codec
+	upgradeHandlers map[string]types.UpgradeHandler
+}
+
+var _ exported.Keeper = (*Keeper)(nil)
+
+// NewKeeper constructs an upgrade Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec) *Keeper {
+	return &Keeper{
+		storeKey:        storeKey,
+		cdc:             cdc,
+		upgradeHandlers: map[string]types.UpgradeHandler{},
+	}
+}
+
+// SetUpgradeHandler sets an UpgradeHandler for the upgrade specified by name. This handler will be called when the upgrade
+// with this name is applied. In order for an upgrade with the given name to proceed, a handler for this upgrade
+// must be set even if it is a no-op function.
+func (k *Keeper) SetUpgradeHandler(name string, upgradeHandler types.UpgradeHandler) {
```

Comment:
Where and when is this called? This is ephemeral state to the keeper. Is that OK?

Reply:
This is set in the app constructor. It probably makes sense to see a real world usage:

https://github.com/regen-network/regen-ledger/pull/69/files#diff-886cebc00c073731af1b3da26f11e536R207-R213